core/lib/spree/core/controller_helpers/order.rb

```diff
@@ -58,7 +58,18 @@ def associate_user
           session[:guest_token] = nil
         end
 
+        def remove_invalid_session_order_id
+          return unless session[:order_id]
+
+          user = try_spree_current_user
+          order = Spree::Order.find(session[:order_id]) rescue nil
+
+          session[:order_id] = nil if order.nil? || order.user != user
```

Comment:
In the case that that the order belongs to a "guest" you have to use `session[:order_token]` to see if the user has access to the current order, so maybe clearing `session[:order_id]` isn't necessary?